### PR TITLE
Change password

### PR DESCRIPTION
**Pull request by Joozty**
This PR adds change password form in settings tab. Password change is handled in an iframe.

### CODE REVIEW

File: components/forms/change-password/index.jsx [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  return (
+    <Card className={className} tag={tag}>
+      <h2>Change password</h2>
+      <iframe
+        ref={iframeRef}
+        className={styles.changePasswordIframe}
+        title="Password change form"
+        src={`/secure/reset.html?${search}`}
+        scrolling="no"
+        allowtransparency="true"
```
Comment (viktor-yakubiv) [inline components/forms/change-password/index.jsx:36]: Why is this needed? It's not standard but the default behaviour is that the iframe is transparent.

File: components/forms/change-password/index.jsx [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+  return (
+    <Card className={className} tag={tag}>
+      <h2>Change password</h2>
+      <iframe
+        ref={iframeRef}
+        className={styles.changePasswordIframe}
+        title="Password change form"
+        src={`/secure/reset.html?${search}`}
+        scrolling="no"
```
Comment (viktor-yakubiv) [inline components/forms/change-password/index.jsx:35]: This attribute is [deprecated](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe#Attributes).

File: components/forms/change-password/index.jsx [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        allowtransparency="true"
+        onLoad={() => {
+          const observer = new MutationObserver(resize)
+          observer.observe(iframeRef.current.contentWindow.document.body, {
+            attributes: true,
+            attributeOldValue: false,
+            characterData: true,
+            characterDataOldValue: false,
+            childList: true,
+            subtree: true,
```
Comment (viktor-yakubiv) [inline components/forms/change-password/index.jsx:45]: Do not observe more than you actually need.

File: components/forms/change-password/index.jsx [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+      <iframe
+        ref={iframeRef}
+        className={styles.changePasswordIframe}
+        title="Password change form"
+        src={`/secure/reset.html?${search}`}
+        scrolling="no"
+        allowtransparency="true"
+        onLoad={() => {
+          const observer = new MutationObserver(resize)
+          observer.observe(iframeRef.current.contentWindow.document.body, {
```
Comment (viktor-yakubiv) [inline components/forms/change-password/index.jsx:39]: Form size never changes. Why do you observe it?

File: components/forms/change-password/styles.css [hunk 5]
```diff
@@ -0,0 +1,5 @@
+.change-password-iframe {
+  width: 100%;
+  max-width: 50rem;
+  height: 100%;
```
Comment (viktor-yakubiv) [inline components/forms/change-password/styles.css:4]: I am not sure `height` property is applied.

File: components/forms/index.js [hunk 6]
```diff
@@ -0,0 +1,2 @@
+// eslint-disable-next-line import/prefer-default-export
+export { default as ChangePassword } from './change-password'
```
Comment (viktor-yakubiv) [inline components/forms/index.js:2]: You could use more modern syntax, we have a plugin for it.

```suggestion
export ChangePassword from './change-password'
```

File: pages/_app/index.jsx [hunk 7]
```diff
@@ -28,11 +28,14 @@ class App extends NextApp {
...
 
+  store = initStore()
+
   handleRouteChange = (url) => {
     logPageView(url)
     const { dataProvider, activity } = new Route(url)
 
-    if (dataProvider) {
+    // eslint-disable-next-line no-restricted-globals
+    if (!isNaN(dataProvider)) {
```
Comment (viktor-yakubiv) [inline pages/_app/index.jsx:38]: What the reason to use global instead Number?

If you ever have to disable linter, always add a comment before with the explanation of the reason for doing so.

File: pages/_app/index.jsx [hunk 8]
```diff
@@ -28,11 +28,14 @@ class App extends NextApp {
     isAuthorized: false,
   }
 
+  store = initStore()
```
Comment (viktor-yakubiv) [inline pages/_app/index.jsx:31]: It executes on the server, does not it? Should it be so?

File: pages/_app/index.jsx [hunk 9]
```diff
@@ -164,12 +163,12 @@ class App extends NextApp {
     return null
   }
 
-  render() {
+  AppShell = () => {
```
Comment (viktor-yakubiv) [inline pages/_app/index.jsx:166]: It seems to be an antipattern to me. Why do you need this?

File: pages/data-providers/[data-provider-id]/settings.css [hunk 10]
```diff
@@ -4,4 +4,8 @@
 
 .section {
   margin-bottom: calc(1.5 * var(--component-padding-y, 1rem));
+
+  & input {
+    max-width: 50rem;
```
Comment (viktor-yakubiv) [inline pages/data-providers/[data-provider-id]/settings.css:9]: You use this value twice. Is not it better to apply this to form as a class and then apply it the iframe.

File: public/secure/reset.js [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+    credentials: 'include',
+    headers: {
+      'Content-Type': 'application/x-www-form-urlencoded',
+    },
+  }).then((response) => {
+    if (response.status === 401)
+      throw new UnauthorisedError('Wrong credentials')
+
+    // TODO: We should distinguish errors.
+    //     Currently API sends only 400 `{"message": "error"}`
```
Comment (viktor-yakubiv) [inline public/secure/reset.js:20]: Yes, but the possible option here is only that the password and password repeat does not match. If something else happened, i.e. email mismatch the token, we should never know about. We just say: "Password reset error. Please request reset password again".

File: components/forms/change-password/index.jsx (hunk 4 above)
Comment (Joozty) [reply components/forms/change-password/index.jsx:39]: It changes when an error/success message is added

File: pages/_app/index.jsx (hunk 8 above)
Comment (Joozty) [reply pages/_app/index.jsx:31]: server has always [new store](https://github.com/oacore/dashboard/blob/master/store/index.jsx#L15). So it should be safe.

File: pages/_app/index.jsx (hunk 9 above)
Comment (Joozty) [reply pages/_app/index.jsx:166]: Actually this commit could be part of https://github.com/oacore/dashboard/pull/263. I need it there because I want to call [store method in Reset component](https://github.com/oacore/dashboard/blob/reset-password/pages/reset/step1.jsx#L20).

File: public/secure/reset.js (hunk 11 above)
Comment (Joozty) [reply public/secure/reset.js:20]: I think 400 is sent instead of 401  when a cookie is not present,i.e. user accidentally logged out. This should be handled differently.

File: pages/_app/index.jsx (hunk 7 above)
Comment (Joozty) [reply pages/_app/index.jsx:38]: Number.isNaN doesn't convert its argument to Number

> true if the given value is NaN and its type is Number; otherwise, false.

I will change it to Number.isNaN(Number(dataProvider))

File: pages/data-providers/[data-provider-id]/settings.css (hunk 10 above)
Comment (Joozty) [reply pages/data-providers/[data-provider-id]/settings.css:9]: it's going to be improved in https://github.com/oacore/dashboard/pull/263. I will leave it as is here for the time being.

File: public/secure/reset.js (hunk 11 above)
Comment (viktor-yakubiv) [reply public/secure/reset.js:26]: You have a good point. @valerapro can you please do changes accordingly?

File: public/secure/reset.js [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+const newPassword = document.getElementById('newPassword')
+const newPasswordAgain = document.getElementById('newPasswordAgain')
+
+function performApiRequest(url, data) {
+  return fetch(url, {
+    method: 'POST',
+    body: data,
+    credentials: 'include',
+    headers: {
+      'Content-Type': 'application/x-www-form-urlencoded',
```
Comment (viktor-yakubiv) [inline public/secure/reset.js:19]: Is not this automatic?

File: public/secure/reset.js [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+  else newPasswordAgain.setCustomValidity('')
+}
+
+function resetInputs() {
+  requiredInputs.forEach((el) => {
+    el.classList.remove('touched')
+  })
+}
+
+function changePassword(event) {
```
Comment (viktor-yakubiv) [inline public/secure/reset.js:60]: When you use const, there is no reason to use `function` unless you want to bind it or need a function prototype.

File: components/forms/change-password/index.jsx [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+    <Card className={className} tag={tag}>
+      <h2>Change password</h2>
+      <iframe
+        ref={iframeRef}
+        className={styles.changePasswordIframe}
+        title="Password change form"
+        src={`/secure/reset.html?${search}`}
+        onLoad={() => {
+          const observer = new ResizeObserver(resize)
+          observer.observe(iframeRef.current.contentWindow.document.body)
```
Comment (viktor-yakubiv) [inline components/forms/change-password/index.jsx:37]: The observer never disconnects from the element. It's better to wrap it into an effect which you can dispose after unmounting.

File: components/forms/change-password/index.jsx [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+    email: store.user?.email,
+  }
+
+  // remove empty params
+  Object.keys(searchParams).forEach(
+    (key) => searchParams[key] === undefined && delete searchParams[key]
+  )
+
+  const search = new URLSearchParams(searchParams).toString()
+  const resize = () => {
```
Comment (viktor-yakubiv) [inline components/forms/change-password/index.jsx:23]: I would use a callback hook but it's minor.

File: components/forms/change-password/index.jsx [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+    iframeRef.current.style.height = `${iframeRef.current.contentWindow.document.body.offsetHeight}px`
```
Comment (viktor-yakubiv) [inline components/forms/change-password/index.jsx:24]: Does not it work without this hack?

File: components/forms/change-password/index.jsx [hunk 17]
```diff
@@ -0,0 +1,44 @@
+import React, { useRef } from 'react'
+
+import styles from './styles.css'
+
+import { Card } from 'design'
+import { withGlobalStore } from 'store'
+
+const { IDP_URL } = process.env
+
+const ChangePassword = ({ className, store, tag = 'section' }) => {
```
Comment (viktor-yakubiv) [inline components/forms/change-password/index.jsx:10]: I do not think makind it being section here is a good idea. You could pass `className` and `tag` as rest props . Also, there is no reason to deny the user pass other HTML attributes.

File: components/forms/change-password/index.jsx [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+  const iframeRef = useRef(null)
+  const searchParams = {
+    identity_provider_url: IDP_URL,
+    email: store.user?.email,
+  }
+
+  // remove empty params
+  Object.keys(searchParams).forEach(
+    (key) => searchParams[key] === undefined && delete searchParams[key]
+  )
```
Comment (viktor-yakubiv) [inline components/forms/change-password/index.jsx:20]: Try to avoid deleting props. It's not clear and does not perform very well.

```suggestion
  const email = store.user?.email
  const searchParams = {
    ...(email != null ? { email } : {}),
    identity_provider_url: IDP_URL,
  }
```

File: pages/data-providers/[data-provider-id]/settings.jsx [hunk 19]
```diff
@@ -98,6 +98,8 @@ const Settings = ({ store, className, ...restProps }) => {
           tag="p"
         />
       </SettingsGroup>
+
+      <ChangePassword className={styles.section} />
```
Comment (viktor-yakubiv) [inline pages/data-providers/[data-provider-id]/settings.jsx:102]: Please, add `tag` here to make it consistent with other.

**Comment by Joozty** [comment]
@viktor-yakubiv these so-called meaningless commits are for you so you don't need to go through all commits again and check if all of your suggestions from commits were fixed. 🙂 Why should I create 10 commits with one line fixes? You won't make me sorry.

File: components/forms/change-password/index.jsx (hunk 16 above)
Comment (Joozty) [reply components/forms/change-password/index.jsx:24]: No, it doesn't...

File: public/secure/reset.js (hunk 12 above)
Comment (Joozty) [reply public/secure/reset.js:19]: Looks like the default value is `Content-Type: application/json` However @valerapro is working on accepting JSON. So I will adjust this code once it's done.

**Comment by viktor-yakubiv** [comment]
I do not mind how you do your commits during the development/review/development but I care what we have in the end. I recommend you to resolve your commits accordingly to the tasks you tried to solve, so in the end, you know that you squash this one with the first, and that one with the second commit. So before merging you can do it and have a well-described job.

"Fixes after review" does not mean anything in the `master` branch and does not mean anything in any squashed commits. I was only to improve existing chunks of job.

File: public/secure/reset.js (hunk 12 above)
Comment (viktor-yakubiv) [reply public/secure/reset.js:19]: It's not. I guess it's empty or `text/plain` by default. Since fetch is a low-level API, it should be empty by default.

Comment (Joozty) [reply public/secure/reset.js:19]: I tried it without and it gave me that header value.

File: components/forms/change-password/index.jsx (hunk 17 above)
Comment (Joozty) [reply components/forms/change-password/index.jsx]: There is no need to pass other HTML attributes. If there is we can simply add it.
